### ISSUE TITLE
admin, customer groups: sql error with Mysql Mode and php warnings x 2

### ISSUE THREAD

**Issue by torvista** [zencart/zencart#4312]
ZC158. mysql-5.7.32, php8, strict reporting, 

1) Development shop
Added new table using sql from upgrade script, in admin SQL patch. No warnings.

On customer groups page load, two php warnings:

> Request URI: /SHOP/ADMIN/index.php?cmd=customer_groups
> IP address: 127.0.0.1
> #1  require(ADMIN/customer_groups.php) called at [ADMIN/index.php:11]
> --> PHP Warning: Undefined variable $gInfo in ADMIN\customer_groups.php on line 213.
> 
> Request URI: /ADMIN/index.php?cmd=customer_groups
> IP address: 127.0.0.1
> #1  require(ADMIN/customer_groups.php) called at [ADMINindex.php:11]
> --> PHP Warning: Attempt to read property "group_id" on null in ADMIN\customer_groups.php on line 213.

2. Vanilla ZC158.
Clean install.  On customer groups pageload, got SQL error. 

> [02-May-2021 19:38:29 Europe/Madrid] Request URI: /zencart-158/admin/index.php?cmd=customer_groups, IP address: 127.0.0.1
#1  trigger_error() called at [...\zencart-158\includes\classes\db\mysql\query_factory.php:657]
#2  queryFactory->show_error() called at [...\zencart-158\includes\classes\db\mysql\query_factory.php:624]
#3  queryFactory->set_error() called at [...\zencart-158\includes\classes\db\mysql\query_factory.php:265]
#4  queryFactory->Execute() called at [...\zencart-158\admin\customer_groups.php:100]
#5  require(...\zencart-158\admin\customer_groups.php) called at [...\zencart-158\admin\index.php:11]
--> PHP Fatal error: 1055:Expression #1 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'zencart-158_local.cg.group_id' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by :: SELECT cg.*, count(ctg.customer_id) as customer_count                          FROM customer_groups cg                         LEFT JOIN customers_to_groups ctg USING (group_id)                         GROUP BY ctg.group_id                          ORDER BY group_name, group_id LIMIT 0, 1 ==> (as called by) ...\zencart-158\admin\customer_groups.php on line 100 <== in ...\zencart-158\includes\classes\db\mysql\query_factory.php on line 657.

On further investigation I find my dev site has define('DB_MYSQL_MODE', 'TRADITIONAL'); in the local configure.

In ZC158 I only find define('DB_MYSQL_MODE', 'TRADITIONAL'); in zc_install\includes\localConfig.php

Admin Version info: Dev Site
MySQL Mode: STRICT_TRANS_TABLES, STRICT_ALL_TABLES, NO_ZERO_IN_DATE, NO_ZERO_DATE, ERROR_FOR_DIVISION_BY_ZERO, TRADITIONAL, NO_AUTO_CREATE_USER, NO_ENGINE_SUBSTITUTION

Admin Version Info: ZC158
MySQL Mode: ONLY_FULL_GROUP_BY, STRICT_TRANS_TABLES, NO_ZERO_IN_DATE, NO_ZERO_DATE, ERROR_FOR_DIVISION_BY_ZERO, NO_AUTO_CREATE_USER, NO_ENGINE_SUBSTITUTION

Both are using the same server.

**Comment by Zen4All** [comment]
Changing the query on line 75 to this:
```php
                $sql = "SELECT cg.*, count(ctg.customer_id) as customer_count 
                        FROM " . TABLE_CUSTOMER_GROUPS . " cg
                        LEFT JOIN " . TABLE_CUSTOMERS_TO_GROUPS . " ctg ON ctg.group_id = cg.group_id
                        GROUP BY cg.group_id 
                        ORDER BY group_name, group_id";
```
worked for me.

**Comment by drbyte** [comment]
With MariaDB 10.5, even with enabling all the possible "strict" settings I can find, I can't trigger this.

That said, I know MySQL has its own idiosyncrasies.

**Comment by Zen4All** [comment]
FYI, this was the error, I got prior to the change
```
[03-May-2021 21:37:42 Europe/Amsterdam] Request URI: /admin/index.php?cmd=customer_groups, IP address: xx.xxx.xx.xx
#1  trigger_error() called at [//public_html/includes/classes/db/mysql/query_factory.php:657]
#2  queryFactory->show_error() called at [//public_html/includes/classes/db/mysql/query_factory.php:624]
#3  queryFactory->set_error() called at [//public_html/includes/classes/db/mysql/query_factory.php:265]
#4  queryFactory->Execute() called at [//admin/customer_groups.php:100]
#5  require(//public_html/admin/customer_groups.php) called at [//public_html/admin/index.php:11]
--> PHP Fatal error: 1055:Expression #1 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'zc158.cg.group_id' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by :: SELECT cg.*, count(ctg.customer_id) as customer_count                          FROM customer_groups cg                         LEFT JOIN customers_to_groups ctg ON cg.group_id = ctg.group_id                         GROUP BY ctg.group_id                          ORDER BY group_name, group_id LIMIT 0, 1 ==> (as called by) //public_html/admin/customer_groups.php on line 100 <== in //public_html/includes/classes/db/mysql/query_factory.php on line 657.

**Comment by drbyte** [comment]
Closed via #4313
